### PR TITLE
FEAT-#4664: Finalize compatibility support for Python 3.6

### LINKED ISSUE #4664: Expand "older pandas" compatibility mode support
As presented by #4301, only a small amount of tests is passing in "compat mode" now:
```
      - run: python -m pytest -n 2 modin/pandas/test/dataframe/test_map_metadata.py
      - run: python -m pytest -n 2 modin/pandas/test/test_series.py
      - run: python -m pytest -n 2 modin/pandas/test/test_general.py
      - run: python -m pytest modin/pandas/test/test_io.py --verbose
```

We should expand the support and at least include `pandas/test/dataframe/test_default.py` and `pandas/test/test_groupby.py` (ideally, cover all the tests to either fix the support or expand the list of known issues).

### PR DESCRIPTION
**Pull request by vnlitvinov**
## What do these changes do?
Run almost all tests from `modin/pandas/test` with only a few cases skipped.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4664
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Review by mvashishtha — COMMENTED** [review]
@vnlitvinov I have some minor comments.

File: modin/_compat/pandas_api/latest/series.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 import numpy as np
 import pandas
+import pandas.core.groupby.generic
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/_compat/pandas_api/latest/series.py:20]: no, that's a leftover from one of my changes, removed

File: modin/_compat/pandas_api/py36/dataframe.py [hunk 2]
```diff
@@ -115,6 +116,35 @@ def prod(
...
+        level=None,
+        fill_value=nan,
+        limit=None,
+        tolerance=None,
+    ):
+        axis = self._get_axis_number(axis)
+        if axis == 0 and labels is not None:
+            index = labels
+        elif labels is not None:
+            columns = labels
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/_compat/pandas_api/py36/dataframe.py:136]: indeed; I've extracted the common code to base class, thanks!

File: modin/pandas/base.py [hunk 3]
```diff
@@ -2284,15 +2285,7 @@ def rmod(
...
-        self,
-        window,
-        min_periods,
-        center,
-        win_type,
-        on,
-        axis,
-        closed,
-        method,
+        self, window, min_periods, center, win_type, *args, **kw
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/pandas/base.py:2288]: I guess this is how I usually specify things; search in Modin codebase shows that usually it's `kwargs`, so I'll fix that

File: modin/pandas/test/dataframe/test_default.py [hunk 4]
```diff
@@ -74,8 +76,20 @@
         ("mask", lambda df: {"cond": df != 0}),
         ("pct_change", None),
         ("to_xarray", None),
-        ("flags", None),
-        ("set_flags", lambda df: {"allows_duplicate_labels": False}),
+        pytest.param(
+            *("flags", None),
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/pandas/test/dataframe/test_default.py:80]: yes 😃 this is to trick `black` to _not_ split the two in two lines 🙃 IMHO in one line it's clearer to see that those two are closely related

File: modin/pandas/test/dataframe/test_default.py [hunk 5]
```diff
@@ -74,8 +76,20 @@
...
+            marks=pytest.mark.skipif(
+                condition=PandasCompatVersion.CURRENT == PandasCompatVersion.PY36,
+                reason="pandas 1.1 does not support .flags",
+            ),
+        ),
+        pytest.param(
+            *("set_flags", lambda df: {"allows_duplicate_labels": False}),
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/pandas/test/dataframe/test_default.py:87]: same, this is to trick `black`

File: modin/pandas/test/dataframe/test_udf.py [hunk 6]
```diff
@@ -79,6 +80,7 @@ def test_agg_apply(axis, func, op):
     eval_general(
         *create_test_dfs(test_data["float_nan_data"]),
         lambda df: getattr(df, op)(func, axis),
+        check_exception_type=PandasCompatVersion.CURRENT != PandasCompatVersion.PY36,
```
[the remark this answers is not in the document]
Comment (vnlitvinov) [reply modin/pandas/test/dataframe/test_udf.py:85]: added

**Comment by vnlitvinov** [comment]
Thanks @mvashishtha, I think I've addressed all your comments, please take another look!

@devin-petersohn @RehanSD 
For reference, here's the set of `groupby`-related tests which are skipped now because older pandas gives weird exceptions from within our implementation. I decided to skip them instead of trying to fix, because they're mostly edge cases related to zero-rows dataframes IIRC.
To see those failures one has to remove skippings in the `test_groupby.py`.
```
========================================================================= short test summary info =======================================================================
FAILED modin/pandas/test/test_groupby.py::test_groupby_with_kwarg_dropna[col_name_multi_by-True] - IndexError: index -1 is out of bounds for axis 0 with size 0
FAILED modin/pandas/test/test_groupby.py::test_groupby_with_kwarg_dropna[col_name+level_name-True] - IndexError: index -1 is out of bounds for axis 0 with size 0
FAILED modin/pandas/test/test_groupby.py::test_groupby_with_kwarg_dropna[col_name-False] - TypeError: 'NoneType' object is not subscriptable
FAILED modin/pandas/test/test_groupby.py::test_groupby_with_kwarg_dropna[col_name-True] - TypeError: Expected tuple, got str
=============================================================== 4 failed, 12 passed, 85 warnings in 54.51s ==============================================================
```

**Review by mvashishtha — DISMISSED** [review]
LGTM @vnlitvinov, thanks for addressing comments!

File: modin/_compat/pandas_api/py36/base.py [hunk 7]
```diff
@@ -29,8 +29,8 @@ class Python36CompatibleBasePandasDataset(BaseCompatibleBasePandasDataset):
     """Compatibility layer for 'Python 3.6 pandas' for Dataset."""
 
     def _validate_ascending(self, ascending):  # noqa: PR01, RT01
```
Comment (YarShev) [inline modin/_compat/pandas_api/py36/base.py:31]: Do we need this method?

File: modin/pandas/base.py (hunk 3 above)
Comment (YarShev) [inline modin/pandas/base.py:2288]:
```suggestion
        self, window, min_periods, center, win_type, *args, **kwargs
```

File: modin/pandas/base.py [hunk 8]
```diff
@@ -2306,10 +2299,8 @@ def _rolling(
                 min_periods=min_periods,
                 center=center,
                 win_type=win_type,
-                on=on,
-                axis=axis,
-                closed=closed,
-                method=method,
+                *args,
+                **kw,
```
Comment (YarShev) [inline modin/pandas/base.py:2303]:
```suggestion
                **kwargs,
```

File: modin/pandas/base.py [hunk 9]
```diff
@@ -2319,10 +2310,8 @@ def _rolling(
             min_periods=min_periods,
             center=center,
             win_type=win_type,
-            on=on,
-            axis=axis,
-            closed=closed,
-            method=method,
+            *args,
+            **kw,
```
Comment (YarShev) [inline modin/pandas/base.py:2314]:
```suggestion
            **kwargs,
```

File: modin/pandas/test/dataframe/test_udf.py [hunk 10]
```diff
@@ -399,4 +406,7 @@ def test_transform(data, func):
     eval_general(
         *create_test_dfs(data),
         lambda df: df.transform(func),
+        # older pandas gives different exception type compared to Modin and current pandas,
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_udf.py:409]: I wonder if the user expect the exception from Modin, which was in older pandas versions, should we raise that exception?

File: modin/pandas/test/test_groupby.py [hunk 11]
```diff
@@ -1305,11 +1375,26 @@ def test_groupby_multiindex(groupby_kwargs):
         pytest.param(
             {"by": ["item0", "one", "two"]},
             id="col_name+level_name",
+            marks=pytest.mark.skipif(
+                PandasCompatVersion.CURRENT == PandasCompatVersion.PY36,
+                reason="Weird exception in Modin internals",
```
Comment (YarShev) [inline modin/pandas/test/test_groupby.py:1380]: Don't we want to file an issue for this?

File: modin/pandas/test/test_groupby.py [hunk 12]
```diff
@@ -1305,11 +1375,26 @@ def test_groupby_multiindex(groupby_kwargs):
...
+            ),
+        ),
+        pytest.param(
+            {"by": ["item0"]},
+            id="col_name",
+            marks=pytest.mark.skipif(
+                PandasCompatVersion.CURRENT == PandasCompatVersion.PY36,
+                reason="Weird exception in Modin internals",
```
Comment (YarShev) [inline modin/pandas/test/test_groupby.py:1388]: same

File: modin/pandas/test/test_groupby.py [hunk 13]
```diff
@@ -1305,11 +1375,26 @@ def test_groupby_multiindex(groupby_kwargs):
...
+            ),
         ),
-        pytest.param({"by": ["item0"]}, id="col_name"),
         pytest.param(
             {"by": ["item0", "item1"]},
             id="col_name_multi_by",
+            marks=pytest.mark.skipif(
+                PandasCompatVersion.CURRENT == PandasCompatVersion.PY36,
+                reason="Weird exception in Modin internals",
```
Comment (YarShev) [inline modin/pandas/test/test_groupby.py:1396]: same

File: modin/pandas/test/test_rolling.py [hunk 14]
```diff
@@ -91,7 +92,10 @@ def test_dataframe(data, window, min_periods, win_type, axis):
         df_equals(modin_rolled.max(), pandas_rolled.max())
         df_equals(modin_rolled.skew(), pandas_rolled.skew())
         df_equals(modin_rolled.kurt(), pandas_rolled.kurt())
-        df_equals(modin_rolled.apply(np.sum), pandas_rolled.apply(np.sum))
+        if PandasCompatVersion.CURRENT != PandasCompatVersion.PY36:
+            # skip this on pandas 1.1 as pandas raise an exception in there
```
Comment (YarShev) [inline modin/pandas/test/test_rolling.py:96]: Don't we want to file an issue for this?

**Review by YarShev — COMMENTED** [review]
The changes are pretty good! Left some comments.

File: modin/_compat/pandas_api/py36/base.py (hunk 7 above)
Comment (vnlitvinov) [reply modin/_compat/pandas_api/py36/base.py:31]: this is for uniform API between different compatibility classes, we'll likely add more in the future...

File: modin/pandas/test/dataframe/test_udf.py (hunk 10 above)
Comment (vnlitvinov) [reply modin/pandas/test/dataframe/test_udf.py:409]: this would complicate our base without much profit; I'd rather not try to support every nook and cranny of older pandas and wait until users explicitly request that support

File: modin/pandas/test/test_groupby.py (hunk 11 above)
Comment (vnlitvinov) [reply modin/pandas/test/test_groupby.py:1380]: Generally, we would, but I skipped the tests which were not working when a groupby resulted in zero-row dataframe, and I think this is were the line should be drawn.

I mean, I don't see a point of submitting an issue which no one is ever planning to fix.

File: modin/pandas/test/test_groupby.py (hunk 12 above)
Comment (vnlitvinov) [reply modin/pandas/test/test_groupby.py:1388]: https://github.com/modin-project/modin/pull/4800#discussion_r944652994

File: modin/pandas/test/test_groupby.py (hunk 13 above)
Comment (vnlitvinov) [reply modin/pandas/test/test_groupby.py:1396]: https://github.com/modin-project/modin/pull/4800#discussion_r944652994

File: modin/pandas/test/test_rolling.py (hunk 14 above)
Comment (vnlitvinov) [reply modin/pandas/test/test_rolling.py:96]: https://github.com/modin-project/modin/pull/4800#discussion_r944652994

File: modin/_compat/pandas_api/py36/base.py (hunk 7 above)
Comment (YarShev) [reply modin/_compat/pandas_api/py36/base.py:31]: I see, thanks!

File: modin/pandas/test/dataframe/test_udf.py (hunk 10 above)
Comment (YarShev) [reply modin/pandas/test/dataframe/test_udf.py:409]: makes sense

File: modin/pandas/test/test_groupby.py (hunk 11 above)
Comment (YarShev) [reply modin/pandas/test/test_groupby.py:1380]: ok, probably the message above makes sense for this too

> wait until users explicitly request that support

**Review by YarShev — APPROVED** [review]
